### PR TITLE
Added `ThrowWithinAsync` assertions

### PR DESCRIPTION
**Pull request by lg2de**
Any thoughts before I continue, @dennisdoomen, @jnyrup?

### CODE REVIEW

**Comment by jnyrup** [comment]
Sorry for the silence.
I would prefer to discuss new APIs in issues before heading to implementations as it makes it easier for us to _only_ focus on the shape of the API and not the implementation details.
Please open an issue and if possible provide some thoughts/examples about the use cases.

This reminds me that I have to finish some documentation on the procress of proposing new APIs.
It will be heavily inspired by https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md, but scaled down to a two-person OSS project.

**Comment by lg2de** [comment]
implements #2034

File: Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt [hunk 1]
```diff
@@ -2273,7 +2275,7 @@ namespace FluentAssertions.Specialized
         protected FluentAssertions.AndConstraint<TAssertions> NotThrowInternal(System.Exception exception, string because, object[] becauseArgs) { }
         protected FluentAssertions.AndConstraint<TAssertions> NotThrowInternal<TException>(System.Exception exception, string because, object[] becauseArgs)
             where TException : System.Exception { }
-        protected FluentAssertions.Specialized.ExceptionAssertions<TException> ThrowInternal<TException>(System.Exception exception, string because, object[] becauseArgs)
+        protected FluentAssertions.Specialized.ExceptionAssertions<TException> ThrowInternal<TException>(System.Exception exception, System.TimeSpan timeSpan, string because, object[] becauseArgs)
```
[the remark this answers is not in the document]
Comment (jnyrup) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt:2278]: We can avoid changing this overload, which I would also prefer since `InvokeWithInterceptionAsync(Func<Task> action, TimeSpan timeSpan)` only uses `action` in the true part and `timeSpan` in the false part.

https://github.com/jnyrup/fluentassertions/commit/9f5676cd7015674d62313a2c1a2239b7e07b6ea5

[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt:2278]: I don't understand what @jnyrup is suggesting, but I would just duplicate the implementation of `ThrowInternal`. In retrospect, it should have never been added to the base-class and should have been a local `private` method. IMO, base-classes should never be used to reuse code. Use composition instead.

[the remark this answers is not in the document]
Comment (jnyrup) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt:2278]: I'm suggesting the same. 
Duplicating `ThrowInternal` and `InvokeWithInterceptionAsync` rather than introducing conditional logic to fit both flows into the same methods.

[the remark this answers is not in the document]
Comment (lg2de) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt:2278]: I'm fine with optimizing the code, but I would like to understand your motivations.

Dennis wrote "IMO, base-classes should **never** be used to reuse code."
Why you are such strict? What is your intention for base-classes?

Jonas wrote "... and timeSpan (only) in the false part" which is not correct because it is part of the condition,
and "... rather than introducing conditional logic to fit both flows into the same methods." without any explanation.

I think having one method `InvokeWithInterceptionAsync` including some variations is easier to understand than two methods with same name and totally different signature and behavior (more than common for method overloads).

[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt:2278]: Let's just say that I've experienced the pain more than a few times. Using base-classes like that leads to classes that don't have a clear purpose other than reusing logic. This often creates unnecessary coupling that I'd like to avoid. I've seen my fair share of those becoming god classes that nobody really understands. In this case, if we would have kept the logic private to the consuming class, we would have never had a need for the 'protected` member in the first place and never had to deal with the potentially breaking change.

[the remark this answers is not in the document]
Comment (lg2de) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt:2278]: Then please define and check such rules for FA globally. 
We have already several base classes with and without protected members.

[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt:2278]: We were discussing this in the context of this specific PR. And those rules are not that black and white. Sometimes a base class is fine and sometimes it isn't. It all depends on its purpose. Also, time will also change the way we look at our own code.

[the remark this answers is not in the document]
Comment (lg2de) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt:2278]: But this makes it complicated for contributors. 
Things I did in the "right way" in the past are "wrong" next time.

[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt:2278]: Yep, that's possible. And that's why we review all PRs ;-)

**Comment by dennisdoomen** [comment]
Now that the API is approved, I guess you can continue, right?

**Comment by lg2de** [comment]
> Now that the API is approved, I guess you can continue, right?

Yes, so far I did not have sufficient time.

File: Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (lg2de) [reply Tests/Approval.Tests/ApprovedApi/FluentAssertions/net47.verified.txt]: I've restarted the PR with all the received input.
Please review again.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 2]
```diff
@@ -167,6 +166,32 @@ public AsyncFunctionAssertions(Func<TTask> subject, IExtractExceptions extractor
...
+    /// <param name="becauseArgs">
+    /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+    /// </param>
+    public async Task<ExceptionAssertions<TException>> ThrowWithinAsync<TException>(
+        TimeSpan timeSpan, string because = "", params object[] becauseArgs)
+        where TException : Exception
+    {
+        Execute.Assertion
+            .ForCondition(Subject is not null)
+            .BecauseOf(because, becauseArgs)
```
Comment (jnyrup) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:187]: We're missing a test that excersizes when `Subject` is null.

File: Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs [hunk 3]
```diff
@@ -49,6 +50,30 @@ protected ExceptionAssertions<TException> ThrowInternal<TException>(Exception ex
...
+        TException[] expectedExceptions = extractor.OfType<TException>(exception).ToArray();
+
+        Execute.Assertion
+            .BecauseOf(because, becauseArgs)
+            .WithExpectation("Expected a <{0}> to be thrown within {1}{reason}, ",
+                typeof(TException), timeSpan)
+            .ForCondition(exception is not null)
+            .FailWith("but no exception was thrown.")
+            .Then
+            .ForCondition(expectedExceptions.Any())
```
Comment (jnyrup) [inline Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs:66]: We're missing a test where the an exception is thrown within the expected timeout, but of an unexpected type.

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 4]
```diff
@@ -180,6 +181,117 @@ public async Task When_task_completes_late_it_should_succeed()
...
+            // Arrange
+            var timer = new FakeClock();
+            var taskFactory = new TaskCompletionSource<bool>();
+
+            // Act
+            Func<Task> action = () =>
+            {
+                return taskFactory
+                    .Awaiting(t => (Task)t.Task)
+                    .Should(timer).ThrowWithinAsync<InvalidOperationException>(100.Milliseconds());
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:284]: This would be a fitting test to `because` and `becauseArgs` and verify them.

**Review by jnyrup — COMMENTED** [review]
Looks really good, thanks!

Only a few comments about test coverage.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs (hunk 2 above)
Comment (lg2de) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:187]: done

File: Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs (hunk 3 above)
Comment (lg2de) [reply Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs:66]: done

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs (hunk 4 above)
Comment (lg2de) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:339]: done

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 5]
```diff
@@ -167,6 +166,32 @@ public AsyncFunctionAssertions(Func<TTask> subject, IExtractExceptions extractor
...
+            .FailWith("Expected {context} to throw {0} within {1}{reason}, but found <null>.",
+                typeof(TException), timeSpan);
+
+        Exception exception = await InvokeWithInterceptionAsync(timeSpan);
```
[the remark this answers is not in the document]
Comment (lg2de) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:205]: done

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 6]
```diff
@@ -177,6 +175,37 @@ public AsyncFunctionAssertions(Func<TTask> subject, IExtractExceptions extractor
...
+        TimeSpan timeSpan, string because = "", params object[] becauseArgs)
+        where TException : Exception
+    {
+        bool success = Execute.Assertion
+            .ForCondition(Subject is not null)
+            .BecauseOf(because, becauseArgs)
+            .FailWith("Expected {context} to throw {0} within {1}{reason}, but found <null>.",
+                typeof(TException), timeSpan);
+
+        if (!success)
```
Comment (jnyrup) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:200]: Let's follow the project style of not using early returns.
(despite our shared personal preference 😉)

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 7]
```diff
@@ -10,7 +10,6 @@ namespace FluentAssertions.Specialized;
 /// <summary>
 /// Contains a number of methods to assert that an asynchronous method yields the expected result.
 /// </summary>
-[DebuggerNonUserCode]
```
Comment (jnyrup) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:13]: revert?

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs (hunk 6 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:200]: > Let's follow the project style of not using early returns.

Early returns at the top of the method are fine by me too. Just not the ones hidden in the middle of the code base.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs (hunk 7 above)
Comment (lg2de) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: ups...

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs (hunk 6 above)
Comment (lg2de) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:201]: Mmh, I just copied from Jonas' changes above...
I'll change it in the same way when discussion in #2105 is completed.

Comment (lg2de) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: done

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 8]
```diff
@@ -180,6 +184,274 @@ public async Task When_task_completes_late_it_should_succeed()
...
+            {
+                return taskFactory
+                    .Awaiting(t => (Task)t.Task)
+                    .Should(timer).ThrowAsync<InvalidOperationException>();
+            };
+            taskFactory.SetException(new InvalidOperationException("foo"));
+
+            // Assert
+            await action.Should().CompleteWithinAsync(1.Seconds());
+        }
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:241]: Disclaimer, I often struggle with mentally keeping track of everything our timing related tests...

But shouldn't we use `timer` for anything? 🤔 
Otherwise, would the snippet below suffice?

```cs
[Fact]
public async Task When_task_throws_it_should_succeed()
{
    // Arrange
    var taskFactory = new TaskCompletionSource<bool>();

    // Act
    Func<Task> action = () =>
    {
        return taskFactory
            .Awaiting(t => (Task)t.Task)
            .Should().ThrowAsync<InvalidOperationException>();
    };
    taskFactory.SetException(new InvalidOperationException("foo"));

    // Assert
    await action.Should().NotThrowAsync();
}
```

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 9]
```diff
@@ -180,6 +184,274 @@ public async Task When_task_completes_late_it_should_succeed()
...
+            {
+                return Awaiting(() => (Task)taskFactory.Task)
+                    .Should(timer).ThrowWithinAsync<InvalidOperationException>(1.Seconds());
+            };
+            var _ = action.Invoke();
+            taskFactory.SetException(new InvalidOperationException("foo"));
+
+            // Assert
+            await action.Should().CompleteWithinAsync(1.Seconds());
+        }
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:362]: Why do we begin executing the task here?
`var _ = action.Invoke();`

That seems odd to me, as if we're not actually testing what we think we're testing.

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs (hunk 8 above)
Comment (lg2de) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:241]: `CompleteWithinAsync` and `NotThrowAsync` has the same meaning but `CompleteWithinAsync` defines a timeout for the test.
But you are right, the timer can be removed here.

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs (hunk 9 above)
Comment (lg2de) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:362]: The test wants to validate the scenario, that the asynchronous method throws AFTER the synchronous portion has been completed. Therefore I need to "invoke".

**Comment by lg2de** [comment]
Please note that I create my commits always to be merged without squashing.
If you disagree with my commits please drop me a note.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 10]
```diff
@@ -366,4 +404,48 @@ private static async Task<Exception> InvokeWithInterceptionAsync(Func<Task> acti
             return exception;
         }
     }
+
+    private async Task<Exception> InvokeWithInterceptionAsync(TimeSpan timeSpan)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:408]: 🔧 This method should be placed directly under the calling `ThrowWithinAsync`
🔧 I think renaming `timeSpan` to `timeOut` makes a bit more self-explanatory

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 11]
```diff
@@ -177,6 +176,37 @@ public AsyncFunctionAssertions(Func<TTask> subject, IExtractExceptions extractor
...
+    {
+        bool success = Execute.Assertion
+            .ForCondition(Subject is not null)
+            .BecauseOf(because, becauseArgs)
+            .FailWith("Expected {context} to throw {0} within {1}{reason}, but found <null>.",
+                typeof(TException), timeSpan);
+
+        if (success)
+        {
+            Exception exception = await InvokeWithInterceptionAsync(timeSpan);
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:203]: 🔧 I think we need to call this variable `caughtException` to emphasize that it can be `null`.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 12]
```diff
@@ -177,6 +176,37 @@ public AsyncFunctionAssertions(Func<TTask> subject, IExtractExceptions extractor
...
+            return ThrowInternal<TException>(exception, timeSpan, because, becauseArgs);
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:204]: 🔧 This method implies it's throwing an exception, but I think the point is to assert that a particular exception was thrown. Let's name it accordingly.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 13]
```diff
@@ -330,12 +361,19 @@ private protected async Task<bool> CompletesWithinTimeoutAsync(Task target, Time
...
+
         if (completedTask != target)
         {
+            // The monitored task did not complete.
             return false;
         }
 
-        // cancel the clock
+        // The monitored task is completed, we shall cancel the clock.
         timeoutCancellationTokenSource.Cancel();
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:377]: 🏕️ This name is confusing. Would you mind renaming it to something that better describes you're canceling the delay?

File: Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs [hunk 14]
```diff
@@ -49,6 +50,30 @@ protected ExceptionAssertions<TException> ThrowInternal<TException>(Exception ex
         return new ExceptionAssertions<TException>(expectedExceptions);
     }
 
+    private protected ExceptionAssertions<TException> ThrowInternal<TException>(
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs:53]: 🔧 This method belongs in the `AsyncFunctionAssertions`,

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs (hunk 10 above)
Comment (IT-VBFK) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:408]: I think `timeout` is the correct casing?

Comment (dennisdoomen) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:408]: Yeah, since we're following the [US spelling](https://grammarist.com/spelling/timeout-time-out/), I guess you're right

Comment (lg2de) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:408]: Ok, done.
So far I supposed ordering by member type ("public", "private", "static", etc.).

File: Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs (hunk 14 above)
Comment (lg2de) [reply Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs:53]: This does not work because it is accessing `private readonly IExtractExceptions extractor`.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs (hunk 10 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: No, we order according to https://csharpcodingguidelines.com/layout-guidelines/#AV2406.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs (hunk 12 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: It's `private protected`, so how would it be a breaking change?

File: Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs (hunk 14 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Specialized/DelegateAssertionsBase.cs:54]: The `AsyncFunctionAssertions` class also has that `IExtractExceptions` as a parameter.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs (hunk 12 above)
Comment (lg2de) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: The new one is private, the old one is protected. The old one was invented and named long time ago.
Both have similar behavior, so both should have similar name, isn't it? 
This is why I recommend consistent renaming with V7.

Comment (dennisdoomen) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: When you move it back to `AsyncFunctionAssertions`, the naming becomes a local concern (as-in within the same class), so it becomes less important the naming is concistent.

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 15]
```diff
@@ -185,6 +187,272 @@ public async Task When_task_completes_late_it_should_succeed()
...
+            var timer = new FakeClock();
+            var taskFactory = new TaskCompletionSource<bool>();
+
+            // Act
+            Func<Task> action = () =>
+            {
+                return Awaiting(() => (Task)taskFactory.Task)
+                    .Should(timer).ThrowWithinAsync<InvalidOperationException>(1.Seconds());
+            };
+            var _ = action.Invoke();
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:358]:
```suggestion
            _ = action.Invoke();
```

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 16]
```diff
@@ -185,6 +187,272 @@ public async Task When_task_completes_late_it_should_succeed()
...
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:446]:
```suggestion
            _ = action.Invoke();
```

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs (hunk 9 above)
Comment (jnyrup) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:363]: Roger.
From CC I see that this also exercises the `return exception;` statement on line 254.

I would have thought we could simplify this scenario with the following, but it fails when for me running the build script.
```cs
var action = async () =>
{
    await Task.Yield();
    throw new InvalidOperationException("foo");
};

// Act
Func<Task> testAction = () => action.Should().ThrowWithinAsync<InvalidOperationException>(1.Seconds());

// Assert
await testAction.Should().NotThrowAsync();
```

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 17]
```diff
@@ -185,6 +187,272 @@ public async Task When_task_completes_late_it_should_succeed()
...
+                    .Should(timer).ThrowWithinAsync<InvalidOperationException>(
+                        100.Milliseconds(), "because we want to test the failure {0}", "message");
+            };
+            timer.Delay(200.Milliseconds());
+
+            // Assert
+            await action.Should().ThrowAsync<XunitException>().WithMessage(
+                "Expected a <System.InvalidOperationException> to be thrown within 100ms"
+                + " because we want to test the failure message,"
+                + " but no exception was thrown.");
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:386]:
```suggestion
            // Arrange
            var timer = new FakeClock();
            var taskFactory = new TaskCompletionSource<bool>();

            // Act
            Func<Task> action = () =>
            {
                return taskFactory
                    .Awaiting(t => (Task)t.Task)
                    .Should(timer).ThrowWithinAsync<InvalidOperationException>(
                        100.Ticks(), "because we want to test the failure {0}", "message");
            };
            timer.Delay(101.Ticks());

            // Assert
            await action.Should().ThrowAsync<XunitException>().WithMessage(
                "Expected a <System.InvalidOperationException> to be thrown within 10.0µs"
                + " because we want to test the failure message,"
                + " but no exception was thrown.");
```

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 18]
```diff
@@ -185,6 +187,272 @@ public async Task When_task_completes_late_it_should_succeed()
...
+            {
+                return taskFactory
+                    .Awaiting(t => (Task)t.Task)
+                    .Should(timer).ThrowWithinAsync<InvalidOperationException>(100.Milliseconds());
+            };
+            taskFactory.SetException(new InvalidOperationException("foo"));
+
+            // Assert
+            await action.Should().CompleteWithinAsync(1.Seconds());
+        }
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:343]: To exercise the edge cases around the `timeSpan` limit.

```cs
[Theory]
[InlineData(99)]
[InlineData(100)]
public async Task When_task_throws_fast_it_should_succeed(int delay)
{
    // Arrange
    var timer = new FakeClock();
    var taskFactory = new TaskCompletionSource<bool>();

    // Act
    Func<Task> action = () =>
    {
        return taskFactory
            .Awaiting(t => (Task)t.Task)
            .Should(timer).ThrowWithinAsync<InvalidOperationException>(100.Ticks());
    };

    timer.Delay(delay.Ticks());
    taskFactory.SetException(new InvalidOperationException("foo"));

    // Assert
    await action.Should().NotThrowAsync();
}

[Fact]
public async Task When_task_throws_slow_it_should_fail()
{
    // Arrange
    var timer = new FakeClock();
    var taskFactory = new TaskCompletionSource<bool>();

    // Act
    Func<Task> action = () =>
    {
        return taskFactory
            .Awaiting(t => (Task)t.Task)
            .Should(timer).ThrowWithinAsync<InvalidOperationException>(100.Ticks());
    };

    timer.Delay(101.Ticks());
    taskFactory.SetException(new InvalidOperationException("foo"));
    timer.Complete();

    // Assert
    await action.Should().ThrowAsync<XunitException>();
}
```

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 19]
```diff
@@ -185,6 +187,272 @@ public async Task When_task_completes_late_it_should_succeed()
...
+                    .Awaiting(t => (Task)t.Task)
+                    .Should(timer).ThrowAsync<InvalidOperationException>();
+            };
+            taskFactory.SetResult(true);
+            timer.Complete();
+
+            // Assert
+            await action.Should().ThrowAsync<XunitException>().WithMessage(
+                "Expected a <System.InvalidOperationException> to be thrown, but no exception was thrown.");
+        }
```
[the remark this answers is not in the document]
Comment (lg2de) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:286]: In the first test you are USING `ThrowAsync` instead of TEST it.
In general you are right and I've updated the tests accordingly.

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs (hunk 18 above)
Comment (lg2de) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:343]: Great work!
The only thing I've changed is using the assertions with timeout.
This is reason why I've introduced them: The tested code may cause deadlocks. In this case the test may never complete.
Asynchronous methods may always be tested with timeout.

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 20]
```diff
@@ -39,12 +42,12 @@ public async Task When_subject_is_null_it_should_throw()
                 timeSpan, "because we want to test the failure {0}", "message");
 
             // Assert
-            await testAction.Should().ThrowAsync<XunitException>()
+            await testAction.Should().ThrowWithinAsync<XunitException>(1.Seconds())
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:45]: Why are you changing many assertions to use `ThrowWithinAsync` and `CompleteWithinAsync`? 🤔

Comment (lg2de) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:45]: See above.

> The only thing I've changed is using the assertions with timeout.
> This is reason why I've introduced them: The tested code may cause deadlocks. In this case the test may never complete.
> Asynchronous methods may always be tested with timeout.

Comment (jnyrup) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:45]: In what case can we deadlock?
Adding timeouts to combat deadlocks sounds like symptomatic treatment to me.

Comment (lg2de) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:45]: Due to bugs in the tested code. This will create early failing tests.

Comment (jnyrup) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:45]: So we're talking about potential deadlocks?
I'd like to not add timeouts prematurely.

Comment (jnyrup) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:45]: For certain test suites it's undoubtedly useful, but let's leave them out of the this test suite.
Our tests are small enough that if we face deadlocks _something_ will fail and we can diagnose it at that time.
For now we think it makes the tests less clear on _exactly_ what scenario they're testing.

Comment (dennisdoomen) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs:45]: I agree with @jnyrup here.

Comment (lg2de) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs]: done

**Review by dennisdoomen — APPROVED** [review]

**Comment by dennisdoomen** [comment]
There's a tiny regression in test coverage. Not sure if it's really an issue.

File: Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs [hunk 21]
```diff
@@ -181,7 +183,286 @@ public async Task When_task_completes_late_it_should_succeed()
...
+            Func<Task> action = () =>
+            {
+                return Awaiting(() => (Task)taskFactory.Task)
+                    .Should(timer).ThrowWithinAsync<InvalidOperationException>(1.Seconds());
+            };
+            _ = action.Invoke();
+            taskFactory.SetException(new NotSupportedException("foo"));
+
+            // Assert
+            await action.Should().ThrowWithinAsync<XunitException>(1.Seconds()).WithMessage(
```
[the remark this answers is not in the document]
Comment (lg2de) [reply Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs]: I've fixed this already.
Maybe you have reviewed the previous version?

**Comment by lg2de** [comment]
> There's a tiny regression in test coverage. Not sure if it's really an issue.

I know that there are lines not covered. But they are edge cases. I tried already to catch them.
I can try again the next days...

**Comment by dennisdoomen** [comment]
> I can try again the next days...

Unless @jnyrup thinks so, don't bother.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 22]
```diff
@@ -180,6 +180,106 @@ public AsyncFunctionAssertions(Func<TTask> subject, IExtractExceptions extractor
...
+                {
+                    // exception in synchronous portion
+                    return result.Exception!.GetBaseException();
+                }
+
+                if (await CompletesWithinTimeoutAsync(result, remainingTime))
+                {
+                    // completed without exception
+                    return null;
+                }
```
Comment (jnyrup) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs:272]: I think we can write it like this to all keep lines covered.
```cs
_ = await CompletesWithinTimeoutAsync(result, remainingTime);

// completed without exception
return null;
```

In the other usages `CompletesWithinTimeoutAsync(` we the Boolean result in the `ForCondition`, whereas here we use the returned exception from this method.

Comment (lg2de) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: You are right, but it's a bit hard explain, WHY we can ignore the result, at least while "quick reading the code"...
So, I tried to write an explanation comment.
Thanks!

**Review by jnyrup — APPROVED** [review]

**Comment by julealgon** [comment]
> Fixes #2034

Looks like this only works in the PR description. The issue is still open.
